diff --git a/node_modules/cashscript/dist/network/ElectrumNetworkProvider.js b/node_modules/cashscript/dist/network/ElectrumNetworkProvider.js
index 5a81f1f..5b335c7 100644
--- a/node_modules/cashscript/dist/network/ElectrumNetworkProvider.js
+++ b/node_modules/cashscript/dist/network/ElectrumNetworkProvider.js
@@ -39,6 +39,7 @@ export default class ElectrumNetworkProvider {
         else if (network === Network.CHIPNET) {
             this.electrum = new ElectrumCluster('CashScript Application', '1.4.1', 1, 1, ClusterOrder.PRIORITY);
             this.electrum.addServer('chipnet.imaginary.cash', 50004, ElectrumTransport.WSS.Scheme, false);
+            this.electrum.addServer('chipnet.bch.ninja', 50004, ElectrumTransport.WSS.Scheme, false);
         }
         else {
             throw new Error(`Tried to instantiate an ElectrumNetworkProvider for unsupported network ${network}`);
